File: devops-scm-api/src/main/java/com/tencent/devops/scm/api/util/GitUtils.java
package com.tencent.devops.scm.api.util;

import static com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_PUSH_ADD_FILE_COUNT;
import static com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_PUSH_ADD_FILE_PREFIX;
import static com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_PUSH_COMMIT_AUTHOR_PREFIX;
import static com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_PUSH_COMMIT_MSG_PREFIX;
import static com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_PUSH_COMMIT_PREFIX;
import static com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_PUSH_COMMIT_TIMESTAMP_PREFIX;
import static com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_PUSH_DELETE_FILE_COUNT;
import static com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_PUSH_DELETE_FILE_PREFIX;
import static com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_PUSH_MODIFY_FILE_COUNT;
import static com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_PUSH_MODIFY_FILE_PREFIX;
import static com.tencent.devops.scm.api.constant.WebhookOutputConstants.MAX_VARIABLE_COUNT;

import com.tencent.devops.scm.api.pojo.Commit;
import org.apache.commons.lang3.StringUtils;

import java.time.ZoneId;
import java.util.HashMap;
import java.util.List;
import java.util.Map;


public class GitUtils {

    public static String expandRef(String name, String prefix) {
        prefix = StringUtils.removeEnd(prefix, "/");
        if (StringUtils.startsWith(name, prefix)) {
            return name;
        }
        return prefix + "/" + name;
    }

    /**
     * 移除分支或tag前缀
     * @param ref 分支或tag
     */
    public static String trimRef(String ref) {
        ref = StringUtils.removeStart(ref, "refs/heads/");
        ref = StringUtils.removeStart(ref, "refs/tags/");
        return ref;
    }

    public static String getShortSha(String commitId) {
        String shortSha = "";
        if (commitId != null && commitId.length() > 8) {
            shortSha = commitId.substring(0, 8);
        }
        return shortSha;
    }

    /**
     * 获取提交变更信息
     * @param commits 提交记录
     * @return
     */
    public static Map<String, Object> getOutputCommitIndexVar(List<Commit> commits) {
        Map<String, Object> startParams = new HashMap<>();
        int addCount = 0;
        int modifyCount = 0;
        int deleteCount = 0;

        for (int index = 0; index < commits.size(); index++) {
            Commit gitCommit = commits.get(index);
            int curIndex = index + 1;
            startParams.put(BK_REPO_GIT_WEBHOOK_PUSH_COMMIT_PREFIX + curIndex, gitCommit.getSha());
            startParams.put(BK_REPO_GIT_WEBHOOK_PUSH_COMMIT_MSG_PREFIX + curIndex, gitCommit.getMessage());
            startParams.put(BK_REPO_GIT_WEBHOOK_PUSH_COMMIT_TIMESTAMP_PREFIX + curIndex,
                    gitCommit.getCommitTime().atZone(ZoneId.systemDefault()).toEpochSecond() * 1000);
            startParams.put(BK_REPO_GIT_WEBHOOK_PUSH_COMMIT_AUTHOR_PREFIX + curIndex, gitCommit.getAuthor());

            addCount += gitCommit.getAdded() != null ? gitCommit.getAdded().size() : 0;
            modifyCount += gitCommit.getModified() != null ? gitCommit.getModified().size() : 0;
            deleteCount += gitCommit.getRemoved() != null ? gitCommit.getRemoved().size() : 0;

            int count = 0;

            if (gitCommit.getAdded() != null) {
                List<String> added = gitCommit.getAdded();
                for (int innerIndex = 0; innerIndex < added.size(); innerIndex++) {
                    startParams.put(BK_REPO_GIT_WEBHOOK_PUSH_ADD_FILE_PREFIX + curIndex + "_" + (innerIndex + 1),
                            added.get(innerIndex));
                    count++;
                    if (count > MAX_VARIABLE_COUNT) {
                        break;
                    }
                }
            }

            if (gitCommit.getModified() != null) {
                List<String> modified = gitCommit.getModified();
                for (int innerIndex = 0; innerIndex < modified.size(); innerIndex++) {
                    startParams.put(BK_REPO_GIT_WEBHOOK_PUSH_MODIFY_FILE_PREFIX + curIndex + "_" + (innerIndex + 1),
                            modified.get(innerIndex));
                    count++;
                    if (count > MAX_VARIABLE_COUNT) {
                        break;
                    }
                }
            }

            if (gitCommit.getRemoved() != null) {
                List<String> removed = gitCommit.getRemoved();
                for (int innerIndex = 0; innerIndex < removed.size(); innerIndex++) {
                    startParams.put(BK_REPO_GIT_WEBHOOK_PUSH_DELETE_FILE_PREFIX + curIndex + "_" + (innerIndex + 1),
                            removed.get(innerIndex));
                    count++;
                    if (count > MAX_VARIABLE_COUNT) {
                        break;
                    }
                }
            }
        }

        startParams.put(BK_REPO_GIT_WEBHOOK_PUSH_ADD_FILE_COUNT, addCount);
        startParams.put(BK_REPO_GIT_WEBHOOK_PUSH_MODIFY_FILE_COUNT, modifyCount);
        startParams.put(BK_REPO_GIT_WEBHOOK_PUSH_DELETE_FILE_COUNT, deleteCount);

        return startParams;
    }
}

File: devops-scm-api/src/main/kotlin/com/tencent/devops/scm/api/util/GitUtils.kt
package com.tencent.devops.scm.api.util

import com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_PUSH_ADD_FILE_COUNT
import com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_PUSH_ADD_FILE_PREFIX
import com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_PUSH_COMMIT_AUTHOR_PREFIX
import com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_PUSH_COMMIT_MSG_PREFIX
import com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_PUSH_COMMIT_PREFIX
import com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_PUSH_COMMIT_TIMESTAMP_PREFIX
import com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_PUSH_DELETE_FILE_COUNT
import com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_PUSH_DELETE_FILE_PREFIX
import com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_PUSH_MODIFY_FILE_COUNT
import com.tencent.devops.scm.api.constant.WebhookOutputCode.BK_REPO_GIT_WEBHOOK_PUSH_MODIFY_FILE_PREFIX
import com.tencent.devops.scm.api.constant.WebhookOutputConstants.MAX_VARIABLE_COUNT
import com.tencent.devops.scm.api.pojo.Commit
import org.apache.commons.lang3.StringUtils
import java.time.ZoneId

/**
 * Git工具类
 */
object GitUtils {
    /**
     * 扩展引用名称
     * @param name 原始名称
     * @param prefix 前缀
     * @return 扩展后的名称
     */
    fun expandRef(name: String, prefix: String): String {
        val trimmedPrefix = StringUtils.removeEnd(prefix, "/")
        return if (StringUtils.startsWith(name, trimmedPrefix)) {
            name
        } else {
            "$trimmedPrefix/$name"
        }
    }

    /**
     * 移除分支或tag前缀
     * @param ref 分支或tag引用
     * @return 处理后的引用
     */
    fun trimRef(ref: String): String {
        var result = StringUtils.removeStart(ref, "refs/heads/")
        result = StringUtils.removeStart(result, "refs/tags/")
        return result
    }

    /**
     * 获取简短的提交SHA
     * @param commitId 完整提交ID
     * @return 前8位字符的提交ID
     */
    fun getShortSha(commitId: String?): String {
        return if (commitId != null && commitId.length > 8) {
            commitId.substring(0, 8)
        } else {
            ""
        }
    }

    /**
     * 获取提交变更信息
     * @param commits 提交记录列表
     * @return 包含提交变更信息的Map
     */
    fun getOutputCommitIndexVar(commits: List<Commit>): Map<String, Any> {
        val startParams = mutableMapOf<String, Any>()
        var addCount = 0
        var modifyCount = 0
        var deleteCount = 0

        commits.forEachIndexed { index, gitCommit ->
            val curIndex = index + 1
            startParams.apply {
                put("$BK_REPO_GIT_WEBHOOK_PUSH_COMMIT_PREFIX$curIndex", gitCommit.sha)
                put("$BK_REPO_GIT_WEBHOOK_PUSH_COMMIT_MSG_PREFIX$curIndex", gitCommit.message)
                put(
                    "$BK_REPO_GIT_WEBHOOK_PUSH_COMMIT_TIMESTAMP_PREFIX$curIndex",
                    gitCommit.commitTime.atZone(ZoneId.systemDefault()).toEpochSecond() * 1000
                )
                put("$BK_REPO_GIT_WEBHOOK_PUSH_COMMIT_AUTHOR_PREFIX$curIndex", gitCommit.author)
            }

            addCount += gitCommit.added.size
            modifyCount += gitCommit.modified.size
            deleteCount += gitCommit.removed.size

            var count = 0

            gitCommit.added.forEachIndexed { innerIndex, file ->
                if (count <= MAX_VARIABLE_COUNT) {
                    startParams["$BK_REPO_GIT_WEBHOOK_PUSH_ADD_FILE_PREFIX${curIndex}_${innerIndex + 1}"] = file
                    count++
                }
            }

            gitCommit.modified.forEachIndexed { innerIndex, file ->
                if (count <= MAX_VARIABLE_COUNT) {
                    startParams["$BK_REPO_GIT_WEBHOOK_PUSH_MODIFY_FILE_PREFIX${curIndex}_${innerIndex + 1}"] = file
                    count++
                }
            }

            gitCommit.removed.forEachIndexed { innerIndex, file ->
                if (count <= MAX_VARIABLE_COUNT) {
                    startParams["$BK_REPO_GIT_WEBHOOK_PUSH_DELETE_FILE_PREFIX${curIndex}_${innerIndex + 1}"] = file
                    count++
                }
            }
        }

        startParams[BK_REPO_GIT_WEBHOOK_PUSH_ADD_FILE_COUNT] = addCount
        startParams[BK_REPO_GIT_WEBHOOK_PUSH_MODIFY_FILE_COUNT] = modifyCount
        startParams[BK_REPO_GIT_WEBHOOK_PUSH_DELETE_FILE_COUNT] =  deleteCount

        return startParams
    }
}
